design/noobaa/system-crd.md

```diff
@@ -0,0 +1,165 @@
+♜ [Rook NooBaa Design](README.md) /
+# System CRD
+
+The operator will define a new kubernetes API group `noobaa.rook.io`. \
+In that group it will define a CRD `System` representing a NooBaa system.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: systems.noobaa.rook.io
+spec:
+  scope: Namespaced
+  group: noobaa.rook.io
+  version: v1alpha1
+  names:
+    kind: System
+    listKind: SystemList
+    singular: system
+    plural: systems
+    shortNames:
+      - nb
+      - noobaa
+```
+
+## Minimal Spec
+
+The operator will provide working defaults when a system created with an empty spec. \
+It assumes the user expects a simple zero-config deployment, with a simple way to configure the system later from the NooBaa UI. \
+This is the minimal system yaml:
+
+```yaml
+apiVersion: noobaa.rook.io/v1alpha1
+kind: System
+metadata:
+  name: noobaa
+  namespace: rook-noobaa
+```
+
+## Full Spec Example
+
+Here is a full blown example of a NooBaa system spec with inline comments.
+
+```yaml
+apiVersion: noobaa.rook.io/v1alpha1
+kind: System
+metadata:
+  name: noobaa-1
+  namespace: rook-noobaa
+spec:
+
+  # Access specifies settings for accessing the system services
+  access:
+
+    # Accounts is a list of named accounts that should get access credentials to NooBaa.
+    accounts:
+      - name: app1-s3-account
+        namespace: app1
+        s3CredentialsSecretName: app1-s3-credentials
+        defaultBackingStoreForNewBuckets: aws-s3
+
+    virtualHost:
+      - virtual.host.name
+
+    security:
+      certificate: TODO
+      disableNonSecureAccess: true
+
+  # Backing stores is a list of storage targets to be used as underlying storage for NooBaa buckets.
+  # NooBaa uses these storage targets to store deduped+compressed+encrypted chunks of data (encryption keys are stored separatly).
+  # Each item will have a locally unique name to identify it when defining bucket placement policies.
+  # Multiple types of storage targets are supported: aws-s3, s3-compatible, google-cloud-storage, azure-blob, obc, pvc.
+  # See specific examples below.
+  backingStores:
```

Comment:
I was wondering about that. Why is that more natural? It is counter-intuitive to me to break this spec down to "top" level entities while these are not top level at all.

Reply:
It's not strictly necessary, and may not even make sense for noobaa, but consider the K8s patterns for other resources. Generally after you create a resource you don't go back and add more subresources to it. If the admin will commonly need to add more resources to the list, it feels like it should be a separate resource. 

Taking the CephCluster as an example, its CRD defines properties related to getting a basic ceph cluster up and running. Depending on your environment, you might create object stores, file systems, or pools on top of that cluster. They depend on the base cluster and require it to be defined first, but their definitions are separate. These could have been defined all in the same CR, but it has been cleaner to be separate CRDs and felt like a more consistent K8s pattern. At the same time, this contributed to the design requirement of one ceph cluster per namespace to keep it simple.

Reply:
Got it.
Adding to your reasons - it seems that separating lists to own CRD's will also allow the OLM integration to be better since each CRD will have its own list in the operator UI.
Making the change and resolving.